hydroflow/tests/surface_codegen.rs

```diff
@@ -114,6 +115,77 @@ pub fn test_large_diamond() {
     df.run_available();
 }
 
+#[test]
+pub fn test_split() {
+    use hydroflow::{tl, tt};
+
+    let (out_send_a, mut out_recv_a) = hydroflow::util::unbounded_channel::<usize>();
+    let (out_send_b, mut out_recv_b) = hydroflow::util::unbounded_channel::<usize>();
+
+    fn split_even_odd(x: usize) -> tt!(Option<usize>, Option<usize>) {
+        if 0 == x % 2 {
+            tl!(Some(x), None)
+        } else {
+            tl!(None, Some(x))
+        }
+    }
+
+    let mut df = hydroflow_syntax! {
+        s = recv_iter(0..10) -> map(split_even_odd) -> split();
+        s[0] -> flatten() -> for_each(|x| out_send_a.send(x).unwrap());
+        s[1] -> flatten() -> for_each(|x| out_send_b.send(x).unwrap());
+    };
+    df.run_available();
+
+    let out_a: Vec<_> = recv_into(&mut out_recv_a);
+    let out_b: Vec<_> = recv_into(&mut out_recv_b);
+    assert_eq!(&[0, 2, 4, 6, 8], &*out_a);
+    assert_eq!(&[1, 3, 5, 7, 9], &*out_b);
+}
+
+#[test]
+pub fn test_split_enum() {
+    #[derive(Split, Debug)]
+    pub enum Value {
+        Integer(u32),
+        String(&'static str),
+        Null,
+        Point { x: f32, y: f32 },
+    }
+
+    let (in_send, in_recv) = hydroflow::util::unbounded_channel::<Value>();
+
+    let (out_send_integer, mut out_recv_integer) = hydroflow::util::unbounded_channel::<u32>();
+    let (out_send_string, mut out_recv_string) =
+        hydroflow::util::unbounded_channel::<&'static str>();
+    let (out_send_null, mut out_recv_null) = hydroflow::util::unbounded_channel::<()>();
+    let (out_send_point, mut out_recv_point) = hydroflow::util::unbounded_channel::<(f32, f32)>();
+
+    let mut df = hydroflow_syntax! {
+        s = recv_stream(in_recv) -> split();
+        s[0] -> flatten() -> for_each(|(x,)| out_send_integer.send(x).unwrap());
```

Comment:
We should definitely have a `stream.attr` syntax at some point...

Which is a bit of a rabbit hole, as then it'd be nice to have named parameters for feeding into operators (which in turn probably would warrant a more function-y syntax for invoking operators, which then would be nice to use with let bindings, and so on...). But would be nice to explore adding such syntactic sugar at some point (unless we want to keep those at a different level and focus on surface syntax as an IR itself).

Reply:
probably just  `a -> [left]my_join` and `my_split[variant] ->` or something